cranelift-faerie/src/backend.rs

```diff
@@ -74,9 +79,164 @@ pub struct FaerieBackend {
     isa: Box<dyn TargetIsa>,
     artifact: faerie::Artifact,
     trap_manifest: Option<FaerieTrapManifest>,
+    frame_sink: Option<FrameSink>,
     libcall_names: Box<dyn Fn(ir::LibCall) -> String>,
 }
 
+struct FrameSink {
+    // we need to retain function names to hand out usize identifiers for FDE addresses,
+    // which are then used to look up function names again for relocations, when `write_address` is
+    // called.
+    fn_names: Vec<String>,
+    table: gimli::write::FrameTable,
+}
+
+impl FrameSink {
+    /// Find a CIE appropriate for the register mapper and initial state provided. This will
+    /// construct a CIE and rely on `gimli` to return an id for an appropriate existing CIE, if
+    /// one exists.
+    ///
+    /// This function also returns a `CFIEncoder` already initialized to the state matching the
+    /// initial CFI instructions for this CIE, ready for use to encode an FDE.
+    pub fn cie_for<'a>(
+        &mut self,
+        initial_state: &[ir::FrameLayoutChange],
+        reg_mapper: &'a DwarfRegMapper,
+    ) -> (gimli::write::CieId, CFIEncoder<'a>) {
+        let mut cie = CommonInformationEntry::new(
+            gimli::Encoding {
+                format: gimli::Format::Dwarf32,
+                version: 1,
+                address_size: 4,
+            },
+            // code alignment factor. Is this right for non-x86_64 ISAs? Probably could be 2 or 4
+            // elsewhere.
+            0x01,
+            // data alignment factor. Same question for non-x86_64 ISAs.
+            -0x08,
```

Comment:
Why negative?

Reply:
I've updated the comment in source to reflect why, but reiterated here: -8 is mostly-arbitrary. I believe on x86 the number _could_ be just about anything, but -8 is what other toolchains use. By using the same parameters in our CIEs, linkers can reuse CIEs and save some space. Other tools picking -8 is likely because it results in data offsets in CFI programs that match stack slots, counting from 0 at the start of the call frame.

Reply:
LLVM uses the stack slot size, and makes it negative if the stack grows down. It doesn't matter as long is it is a factor, and matching other tools makes sense.